### PR TITLE
Av1 tiles

### PR DESCRIPTION
**Pull request by chkngit**
av1 tiles functionality
usage: 
 -tile-row x
 -tile-col y
Log 2 Tile Rows and columns . 0 means no tiling,1 means that we split the dimension into 2

### CODE REVIEW

File: Source/Lib/Codec/EbDefinitions.h [hunk 1]
```diff
@@ -46,6 +46,7 @@ extern "C" {
 #define REST_M        1 // multi-threaded restoration
 #define REST_NEED_B   1 // use boundary update in restoration
 #define NEW_PRED_STRUCT                                 1 // Ability to run 5-layer prediction structure. By Default 5L is used
+#define  TILES     1
```
Comment (joelsole) [inline Source/Lib/Codec/EbDefinitions.h]: Is it necessary to add the #define, and the corresponding code duplication?

File: Source/Lib/Codec/EbEntropyCoding.c [hunk 2]
```diff
@@ -350,15 +350,23 @@ void GetTxbCtx(
 
     uint8_t sign;
 
+#if TILES
+    if (dcSignLevelCoeffNeighborArray->topArray[dcSignLevelCoeffTopNeighborIndex] != INVALID_NEIGHBOR_DATA){
```
Comment (joelsole) [inline Source/Lib/Codec/EbEntropyCoding.c]: dcSignLevelCoeffNeighborArray, topArray => snake_case

File: Source/Lib/Codec/EbEntropyCoding.c [hunk 3]
```diff
@@ -3006,6 +3031,175 @@ void av1_get_tile_limits(PictureParentControlSet_t * pcsPtr) {
     cm->min_log2_tiles = AOMMAX(cm->min_log2_tiles, cm->min_log2_tile_cols);
 }
 
+#if TILES
+void av1_calculate_tile_cols(PictureParentControlSet_t * pcsPtr) {
```
Comment (joelsole) [inline Source/Lib/Codec/EbEntropyCoding.c]: pcsPtr => snake_case

File: Source/Lib/Codec/EbEntropyCoding.c [hunk 4]
```diff
@@ -3006,6 +3031,175 @@ void av1_get_tile_limits(PictureParentControlSet_t * pcsPtr) {
...
+    }
+    else {
+        int max_tile_area_sb = (sb_rows * sb_cols);
+        int widest_tile_sb = 1;
+        cm->log2_tile_cols = tile_log2(1, cm->tile_cols);
+        for (i = 0; i < cm->tile_cols; i++) {
+            int size_sb = cm->tile_col_start_sb[i + 1] - cm->tile_col_start_sb[i];
+            widest_tile_sb = AOMMAX(widest_tile_sb, size_sb);
+        }
+        if (cm->min_log2_tiles) {
```
Comment (joelsole) [inline Source/Lib/Codec/EbEntropyCoding.c]: braces are only necessary for multi-line code blocks or multi-line condition statements;

File: Source/Lib/Codec/EbEntropyCoding.c [hunk 5]
```diff
@@ -3006,6 +3031,175 @@ void av1_get_tile_limits(PictureParentControlSet_t * pcsPtr) {
...
+
+        cm->tile_height = size_sb << pcsPtr->sequence_control_set_ptr->mib_size_log2;
+        cm->tile_height = AOMMIN(cm->tile_height, cm->mi_rows);
+    }
+    else {
+        cm->log2_tile_rows = tile_log2(1, cm->tile_rows);
+    }
+}
+
+ void set_tile_info(PictureParentControlSet_t * pcsPtr)
```
Comment (joelsole) [inline Source/Lib/Codec/EbEntropyCoding.c]: snake case

File: Source/Lib/Codec/EbEncHandle.c [hunk 6]
```diff
@@ -2485,6 +2489,12 @@ static EbErrorType VerifySettings(
         SVT_LOG("Error Instance %u: The lookahead distance must be [0 - %d] \n", channelNumber + 1, MAX_LAD);
         return_error = EB_ErrorBadParameter;
     }
+#if TILES
+    if (config->tile_rows > 6 || config->tile_columns > 6) {
```
Comment (joelsole) [inline Source/Lib/Codec/EbEncHandle.c]: shall config->tile_rows < 0 || config->tile_columns < 0  also be included?

File: Source/Lib/Codec/EbDefinitions.h (hunk 1 above)
Comment (chkngit) [reply Source/Lib/Codec/EbDefinitions.h]: it is a good practice to have macro that we have found very helpful. two advantages: first you can localize all changes easily by searching the FLAG, and Second you can turn OFF the changes completely or even better some parts of the changes when there are problems. We usually keep the FLAG for few weeks then we remove them once the code is settled and seems stable enough

File: Source/Lib/Codec/EbEntropyCoding.c (hunk 2 above)
Comment (chkngit) [reply Source/Lib/Codec/EbEntropyCoding.c]: needs to touch lot of non formatted code, which will be done in the next code clean up phase.

File: Source/Lib/Codec/EbEntropyCoding.c (hunk 3 above)
Comment (chkngit) [reply Source/Lib/Codec/EbEntropyCoding.c]: ok, will fix this one

File: Source/Lib/Codec/EbEntropyCoding.c (hunk 4 above)
Comment (chkngit) [reply Source/Lib/Codec/EbEntropyCoding.c]: ok, will fix this one

File: Source/Lib/Codec/EbEntropyCoding.c (hunk 5 above)
Comment (chkngit) [reply Source/Lib/Codec/EbEntropyCoding.c]: ok, will fix this one

File: Source/Lib/Codec/EbEncHandle.c (hunk 6 above)
Comment (chkngit) [reply Source/Lib/Codec/EbEncHandle.c]: ok, make sense

**Comment by EwoutH** [comment]
Can we use `-tile-columns` and `-tile-rows`? That way we keep it the same as aomenc.

**Comment by chkngit** [comment]
change config parameters to -tile-rows and -tile-columns to match aom

File: Config/Sample.cfg [hunk 7]
```diff
@@ -42,6 +42,10 @@ SceneChangeDetection            : 0             # Enable Scene Change Detection
 LookAheadDistance               : 17            # Number of picture lookahead (0: no lookahead) [0-120]
 ImproveSharpness                : 0             # Improve sharpness (0= OFF, 1=ON )
 
+#====================== Tiles ===============================
+TileRow              : 1             # log2 Tile Rows
```
Comment (hassount) [inline Config/Sample.cfg]: can we add a range of parameters here ? what's the maximum value for the rows and columns

Comment (chkngit) [reply Config/Sample.cfg]: ok.

**Pull request merged**